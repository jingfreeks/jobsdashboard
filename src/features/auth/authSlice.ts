import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';

interface AuthState {
  user: null | { email: string; name?: string };
  loading: boolean;
  error: string | null;
}

const initialState: AuthState = {
  user: null,
  loading: false,
  error: null,
};

const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    loginStart(state) {
      state.loading = true;
      state.error = null;
    },
    loginSuccess(state, action: PayloadAction<{ email: string; name?: string }>) {
      state.user = action.payload;
      state.loading = false;
      state.error = null;
    },
    loginFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.error = action.payload;
    },
    registerStart(state) {
      state.loading = true;
      state.error = null;
    },
    registerSuccess(state, action: PayloadAction<{ email: string; name?: string }>) {
      state.user = action.payload;
      state.loading = false;
      state.error = null;
    },
    registerFailure(state, action: PayloadAction<string>) {
      state.loading = false;
      state.error = action.payload;
    },
    logout(state) {
      state.user = null;
      state.loading = false;
      state.error = null;
    },
  },
});

export const {
  loginStart,
  loginSuccess,
  loginFailure,
  registerStart,
  registerSuccess,
  registerFailure,
  logout,
} = authSlice.actions;

export default authSlice.reducer; 