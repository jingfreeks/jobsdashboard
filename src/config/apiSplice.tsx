/* eslint-disable @typescript-eslint/no-explicit-any */
import {setCredentials, setLogout} from '@/features/auth';
import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
const baseUrl = import.meta.env.VITE_API_BASE_URL;
const baseQuery = fetchBaseQuery({
  baseUrl,
  credentials: 'include',
   
  prepareHeaders: (headers, {getState}: any) => {
    const token = getState().auth.token;
    if (token) {
      headers.set('authorization', `Bearer ${token}`);
    }
    return headers;
  },
});

const baseQueryWithAuth = async (
  args: string,
  api: any,
  extraOptions: string,
) => {
  let result: any = await baseQuery(args, api, extraOptions);
  console.log(result);
  if (result?.error?.originalStatus === 403) {
    const refereshResult = await baseQuery('/refresh', api, extraOptions);

    if (refereshResult?.data) {
      const currentAuth = api.getState().auth;
      //store the new token
      api.dispatch(setCredentials({
        user: currentAuth.user,
        accessToken: (refereshResult.data as any).accessToken,
        userId: currentAuth.userId,
        roles: currentAuth.roles
      }));
      //retry the original query with new access token
      result = await baseQuery(args, api, extraOptions);
    } else {
      api.dispatch(setLogout());
    }
  }
  return result;
};

export const apiSlice = createApi<any, any>({
  baseQuery: baseQueryWithAuth,
  reducerPath: 'api',
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  endpoints: _ => ({}),
  tagTypes: ['City', 'Jobs', 'MyJobs', 'JobDetails', 'Profile'] as any,
});