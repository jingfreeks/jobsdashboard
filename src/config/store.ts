import { configureStore } from '@reduxjs/toolkit';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from '@/features/auth';
import {apiSlice} from './apiSplice';

// Configure persistence for auth reducer
const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['user', 'token', 'userId', 'roles'], // Only persist these fields
  timeout: 10000, // 10 second timeout for storage operations
  serialize: true, // Enable serialization
  deserialize: true, // Enable deserialization
};

const persistedAuthReducer = persistReducer(authPersistConfig, authReducer);

export const store = configureStore({
  reducer: {
    auth: persistedAuthReducer,
    [apiSlice.reducerPath]: apiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: ['auth'], // Ignore auth state for serialization checks
      },
      immutableCheck: {
        ignoredPaths: ['auth'], // Ignore auth state for immutability checks
      },
    }).concat(apiSlice.middleware),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch; 