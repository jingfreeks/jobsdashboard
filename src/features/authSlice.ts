import {apiSlice} from '@/config/apiSplice';

export const authApiSlice = apiSlice.injectEndpoints({
  endpoints: builder => ({
    login: builder.mutation({
      query: credentials => ({
        url: '/auth',
        method: 'POST',
        body: {...credentials},
      }),
    }),
    logout: builder.mutation({
      query: () => ({
        url: '/logout',
        method: 'POST',
        // body: {...credentials},
      }),
    }),
    signup: builder.mutation({
      query: credentials => ({
        url: '/signup',
        method: 'POST',
        body: {...credentials},
      }),
    }),
  }),
  overrideExisting: true,
});

export const {useLoginMutation,useSignupMutation,useLogoutMutation} = authApiSlice || {};