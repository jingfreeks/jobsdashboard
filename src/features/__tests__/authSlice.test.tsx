import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { apiSlice } from '@/config/apiSplice';
import { authApiSlice, useLoginMutation, useSignupMutation, useLogoutMutation } from '../authSlice';

// Mock fetch for API calls
const mockFetch = vi.fn();
global.fetch = mockFetch;

describe('Auth API Slice', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('Auth API Slice Structure', () => {
    it('should have authApiSlice defined', () => {
      expect(authApiSlice).toBeDefined();
    });

    it('should match snapshot for auth API slice', () => {
      expect(authApiSlice).toMatchSnapshot();
    });

    it('should have endpoints property', () => {
      expect(authApiSlice.endpoints).toBeDefined();
    });

    it('should have login endpoint', () => {
      expect(authApiSlice.endpoints.login).toBeDefined();
    });

    it('should have logout endpoint', () => {
      expect(authApiSlice.endpoints.logout).toBeDefined();
    });

    it('should have signup endpoint', () => {
      expect(authApiSlice.endpoints.signup).toBeDefined();
    });

    it('should have correct endpoint configurations', () => {
      // Test login endpoint
      const loginEndpoint = authApiSlice.endpoints.login;
      expect(loginEndpoint.name).toBe('login');
      
      // Test logout endpoint
      const logoutEndpoint = authApiSlice.endpoints.logout;
      expect(logoutEndpoint.name).toBe('logout');
      
      // Test signup endpoint
      const signupEndpoint = authApiSlice.endpoints.signup;
      expect(signupEndpoint.name).toBe('signup');
    });
  });

  describe('Exported Hooks', () => {
    it('should export useLoginMutation', () => {
      expect(useLoginMutation).toBeDefined();
      expect(typeof useLoginMutation).toBe('function');
    });

    it('should export useSignupMutation', () => {
      expect(useSignupMutation).toBeDefined();
      expect(typeof useSignupMutation).toBe('function');
    });

    it('should export useLogoutMutation', () => {
      expect(useLogoutMutation).toBeDefined();
      expect(typeof useLogoutMutation).toBe('function');
    });

    it('should match snapshot for exported hooks', () => {
      const hooks = {
        useLoginMutation: typeof useLoginMutation,
        useSignupMutation: typeof useSignupMutation,
        useLogoutMutation: typeof useLogoutMutation,
      };
      expect(hooks).toMatchSnapshot();
    });
  });

  describe('API Endpoint Configuration', () => {
    it('should have correct login endpoint configuration', () => {
      const loginEndpoint = authApiSlice.endpoints.login;
      
      expect(loginEndpoint.name).toBe('login');
    });

    it('should have correct signup endpoint configuration', () => {
      const signupEndpoint = authApiSlice.endpoints.signup;
      
      expect(signupEndpoint.name).toBe('signup');
    });

    it('should have correct logout endpoint configuration', () => {
      const logoutEndpoint = authApiSlice.endpoints.logout;
      
      expect(logoutEndpoint.name).toBe('logout');
    });

    it('should match snapshot for endpoint configurations', () => {
      const endpointConfigs = {
        login: {
          name: authApiSlice.endpoints.login.name,
        },
        signup: {
          name: authApiSlice.endpoints.signup.name,
        },
        logout: {
          name: authApiSlice.endpoints.logout.name,
        },
      };

      expect(endpointConfigs).toMatchSnapshot();
    });
  });

  describe('API Slice Properties', () => {
    it('should have reducer property', () => {
      expect(authApiSlice.reducer).toBeDefined();
      expect(typeof authApiSlice.reducer).toBe('function');
    });

    it('should have middleware property', () => {
      expect(authApiSlice.middleware).toBeDefined();
      expect(typeof authApiSlice.middleware).toBe('function');
    });

    it('should have reducerPath property', () => {
      expect(authApiSlice.reducerPath).toBeDefined();
      expect(typeof authApiSlice.reducerPath).toBe('string');
    });

    it('should have internalActions property', () => {
      expect(authApiSlice.internalActions).toBeDefined();
      expect(typeof authApiSlice.internalActions).toBe('object');
    });

    it('should have util property', () => {
      expect(authApiSlice.util).toBeDefined();
      expect(typeof authApiSlice.util).toBe('object');
    });

    it('should match snapshot for API slice properties', () => {
      const properties = {
        hasReducer: !!authApiSlice.reducer,
        hasMiddleware: !!authApiSlice.middleware,
        reducerPath: authApiSlice.reducerPath,
        hasInternalActions: !!authApiSlice.internalActions,
        hasUtil: !!authApiSlice.util,
      };

      expect(properties).toMatchSnapshot();
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle undefined endpoints gracefully', () => {
      // Test that accessing non-existent endpoints doesn't crash
      expect(() => {
        const endpoints = authApiSlice.endpoints;
        expect(endpoints).toBeDefined();
      }).not.toThrow();
    });

    it('should have consistent endpoint structure', () => {
      const loginEndpoint = authApiSlice.endpoints.login;
      const signupEndpoint = authApiSlice.endpoints.signup;
      const logoutEndpoint = authApiSlice.endpoints.logout;

      // All endpoints should have the same basic structure
      const expectedProperties = ['name', 'initiate', 'select', 'useMutation'];
      
      expectedProperties.forEach(prop => {
        expect(loginEndpoint).toHaveProperty(prop);
        expect(signupEndpoint).toHaveProperty(prop);
        expect(logoutEndpoint).toHaveProperty(prop);
      });
    });

    it('should have unique endpoint names', () => {
      const loginName = authApiSlice.endpoints.login.name;
      const signupName = authApiSlice.endpoints.signup.name;
      const logoutName = authApiSlice.endpoints.logout.name;

      expect(loginName).not.toBe(signupName);
      expect(loginName).not.toBe(logoutName);
      expect(signupName).not.toBe(logoutName);
    });
  });

  describe('Performance and Memory', () => {
    it('should have efficient endpoint access', () => {
      // Test that accessing endpoints multiple times doesn't cause issues
      for (let i = 0; i < 100; i++) {
        expect(authApiSlice.endpoints.login).toBeDefined();
        expect(authApiSlice.endpoints.signup).toBeDefined();
        expect(authApiSlice.endpoints.logout).toBeDefined();
      }
    });

    it('should maintain consistent references', () => {
      const loginEndpoint1 = authApiSlice.endpoints.login;
      const loginEndpoint2 = authApiSlice.endpoints.login;
      
      expect(loginEndpoint1).toBe(loginEndpoint2);
    });

    it('should handle rapid property access', () => {
      const properties = ['reducer', 'middleware', 'reducerPath', 'endpoints', 'util'];
      
      for (let i = 0; i < 100; i++) {
        properties.forEach(prop => {
          expect(authApiSlice).toHaveProperty(prop);
        });
      }
    });
  });

  describe('Enhanced Hook Testing', () => {
    let store: ReturnType<typeof createTestStore>;

    const createTestStore = () => {
      return configureStore({
        reducer: {
          api: apiSlice.reducer,
        },
        middleware: (getDefaultMiddleware) =>
          getDefaultMiddleware().concat(apiSlice.middleware),
      });
    };

    beforeEach(() => {
      store = createTestStore();
      vi.clearAllMocks();
      mockFetch.mockClear();
    });

    afterEach(() => {
      vi.restoreAllMocks();
    });

    const TestWrapper = ({ children }: { children: React.ReactNode }) => (
      <Provider store={store}>{children}</Provider>
    );

    describe('Login Mutation Hook', () => {
      it('should have working login mutation hook with full structure', () => {
        const { result } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

        expect(result.current[0]).toBeDefined();
        expect(result.current[1]).toBeDefined();
        expect(typeof result.current[0]).toBe('function');
        expect(typeof result.current[1].isLoading).toBe('boolean');
        expect(typeof result.current[1].isError).toBe('boolean');
        expect(typeof result.current[1].isSuccess).toBe('boolean');
        expect(typeof result.current[1].reset).toBe('function');
      });

      it('should handle login mutation with various credential types', () => {
        const { result } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

        const [login] = result.current;

        // Test that the mutation function can be called with different credential types
        expect(typeof login).toBe('function');
        
        // Test with empty credentials
        expect(() => login({})).not.toThrow();
        
        // Test with basic credentials
        expect(() => login({ email: 'test@example.com', password: 'password' })).not.toThrow();
        
        // Test with extended credentials
        expect(() => login({ 
          email: 'test@example.com', 
          password: 'password',
          rememberMe: true 
        })).not.toThrow();
      });
    });

    describe('Signup Mutation Hook', () => {
      it('should have working signup mutation hook with full structure', () => {
        const { result } = renderHook(() => useSignupMutation(), {
          wrapper: TestWrapper,
        });

        expect(result.current[0]).toBeDefined();
        expect(result.current[1]).toBeDefined();
        expect(typeof result.current[0]).toBe('function');
        expect(typeof result.current[1].isLoading).toBe('boolean');
        expect(typeof result.current[1].isError).toBe('boolean');
        expect(typeof result.current[1].isSuccess).toBe('boolean');
        expect(typeof result.current[1].reset).toBe('function');
      });

      it('should handle signup mutation with various credential types', () => {
        const { result } = renderHook(() => useSignupMutation(), {
          wrapper: TestWrapper,
        });

        const [signup] = result.current;

        // Test that the mutation function can be called with different credential types
        expect(typeof signup).toBe('function');
        
        // Test with minimal credentials
        expect(() => signup({ email: 'test@example.com', password: 'password' })).not.toThrow();
        
        // Test with extended credentials
        expect(() => signup({ 
          email: 'test@example.com', 
          password: 'password',
          name: 'Test User',
          phone: '+1234567890'
        })).not.toThrow();
        
        // Test with complex credentials
        expect(() => signup({ 
          email: 'test@example.com', 
          password: 'password',
          name: 'Test User',
          preferences: { theme: 'dark', notifications: true }
        })).not.toThrow();
      });
    });

    describe('Logout Mutation Hook', () => {
      it('should have working logout mutation hook with full structure', () => {
        const { result } = renderHook(() => useLogoutMutation(), {
          wrapper: TestWrapper,
        });

        expect(result.current[0]).toBeDefined();
        expect(result.current[1]).toBeDefined();
        expect(typeof result.current[0]).toBe('function');
        expect(typeof result.current[1].isLoading).toBe('boolean');
        expect(typeof result.current[1].isError).toBe('boolean');
        expect(typeof result.current[1].isSuccess).toBe('boolean');
        expect(typeof result.current[1].reset).toBe('function');
      });

      it('should handle logout mutation with various parameter types', () => {
        const { result } = renderHook(() => useLogoutMutation(), {
          wrapper: TestWrapper,
        });

        const [logout] = result.current;

        // Test that the mutation function can be called with different parameter types
        expect(typeof logout).toBe('function');
        
                 // Test with no parameters
         expect(() => logout(undefined)).not.toThrow();
        
        // Test with empty object
        expect(() => logout({})).not.toThrow();
        
        // Test with null
        expect(() => logout(null)).not.toThrow();
      });
    });

    describe('Hook State Management', () => {
      it('should handle multiple hook instances', () => {
        const { result: loginResult } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

        const { result: signupResult } = renderHook(() => useSignupMutation(), {
          wrapper: TestWrapper,
        });

        const { result: logoutResult } = renderHook(() => useLogoutMutation(), {
          wrapper: TestWrapper,
        });

        // All hooks should be independent
        expect(loginResult.current[0]).not.toBe(signupResult.current[0]);
        expect(loginResult.current[0]).not.toBe(logoutResult.current[0]);
        expect(signupResult.current[0]).not.toBe(logoutResult.current[0]);
      });

      it('should handle hook state transitions', () => {
        const { result } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

                 const [, state] = result.current;

                 // Initial state should be idle
        expect(state.isLoading).toBe(false);
        expect(state.isError).toBe(false);
        expect(state.isSuccess).toBe(false);
        expect(state.data).toBeUndefined();
        expect(state.error).toBeUndefined();

         // Test reset function
         expect(() => state.reset()).not.toThrow();
      });
    });

    describe('Hook Error Handling', () => {
      it('should handle hook errors gracefully', () => {
        const { result } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

                 const [, state] = result.current;

                 // Test that error states are properly typed
        expect(state.error).toBeUndefined();
        expect(state.isError).toBe(false);

         // Test reset functionality
         expect(typeof state.reset).toBe('function');
         expect(() => state.reset()).not.toThrow();
      });
    });

    describe('Hook Performance', () => {
      it('should handle rapid hook calls', () => {
        const { result } = renderHook(() => useLoginMutation(), {
          wrapper: TestWrapper,
        });

        const [login] = result.current;

        // Test rapid calls to the mutation function
        for (let i = 0; i < 10; i++) {
          expect(() => login({ 
            email: `user${i}@example.com`, 
            password: 'password' 
          })).not.toThrow();
        }
      });

      it('should maintain hook consistency under load', () => {
        const { result } = renderHook(() => useSignupMutation(), {
          wrapper: TestWrapper,
        });

        const [signup, state] = result.current;

        // Test multiple rapid calls
        for (let i = 0; i < 5; i++) {
          expect(() => signup({ 
            email: `user${i}@example.com`, 
            password: 'password',
            name: `User ${i}`
          })).not.toThrow();
        }

        // State should remain consistent
        expect(typeof state.isLoading).toBe('boolean');
        expect(typeof state.isError).toBe('boolean');
        expect(typeof state.isSuccess).toBe('boolean');
      });
    });
  });
}); 