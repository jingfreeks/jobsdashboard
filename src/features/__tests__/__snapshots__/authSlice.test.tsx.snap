// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auth API Slice > API Endpoint Configuration > should match snapshot for endpoint configurations 1`] = `
{
  "login": {
    "name": "login",
  },
  "logout": {
    "name": "logout",
  },
  "signup": {
    "name": "signup",
  },
}
`;

exports[`Auth API Slice > API Slice Properties > should match snapshot for API slice properties 1`] = `
{
  "hasInternalActions": true,
  "hasMiddleware": true,
  "hasReducer": true,
  "hasUtil": true,
  "reducerPath": "api",
}
`;

exports[`Auth API Slice > Auth API Slice Structure > should match snapshot for auth API slice 1`] = `
{
  "endpoints": {
    "login": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "login",
      "select": [Function],
      "useMutation": [Function],
    },
    "logout": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "logout",
      "select": [Function],
      "useMutation": [Function],
    },
    "signup": {
      "initiate": [Function],
      "matchFulfilled": [Function],
      "matchPending": [Function],
      "matchRejected": [Function],
      "name": "signup",
      "select": [Function],
      "useMutation": [Function],
    },
  },
  "enhanceEndpoints": [Function],
  "injectEndpoints": [Function],
  "internalActions": {
    "cacheEntriesUpserted": [Function],
    "internal_getRTKQSubscriptions": [Function],
    "middlewareRegistered": [Function],
    "onFocus": [Function],
    "onFocusLost": [Function],
    "onOffline": [Function],
    "onOnline": [Function],
    "queryResultPatched": [Function],
    "removeMutationResult": [Function],
    "removeQueryResult": [Function],
    "resetApiState": [Function],
    "subscriptionsUpdated": [Function],
    "unsubscribeQueryResult": [Function],
    "updateProvidedBy": [Function],
    "updateSubscriptionOptions": [Function],
  },
  "middleware": [Function],
  "reducer": [Function],
  "reducerPath": "api",
  "useLoginMutation": [Function],
  "useLogoutMutation": [Function],
  "usePrefetch": [Function],
  "useSignupMutation": [Function],
  "util": {
    "getRunningMutationThunk": [Function],
    "getRunningMutationsThunk": [Function],
    "getRunningQueriesThunk": [Function],
    "getRunningQueryThunk": [Function],
    "invalidateTags": [Function],
    "patchQueryData": [Function],
    "prefetch": [Function],
    "resetApiState": [Function],
    "selectCachedArgsForQuery": [Function],
    "selectInvalidatedBy": [Function],
    "updateQueryData": [Function],
    "upsertQueryData": [Function],
    "upsertQueryEntries": [Function],
  },
}
`;

exports[`Auth API Slice > Exported Hooks > should match snapshot for exported hooks 1`] = `
{
  "useLoginMutation": "function",
  "useLogoutMutation": "function",
  "useSignupMutation": "function",
}
`;
